bot-settings: Add modal to generate Integration URL.

sbansal1999 opened the pull request:
This PR adds a modal that makes it easier to generate an Integration URL. The original issue also says about creating a checkbox UI for adding events that will be used to filter notifications -- I think that can be added in a follow-up PR.

The modal can be accessed through the `Bots` section in `Personal Settings` and also from the `Edit bot` menu for 
`Incoming webhook` bots only as it doesn't make sense to create Integration URLs for other types of bots.

To access the names of all the integration -- `display_name` and `all_event_types` fields were added in the `realm_incoming_webhook_bots` object in `page_params`. [Link to the discussion in #api-design](https://chat.zulip.org/#narrow/stream/378-api-design/topic/integration.20display.20name)

Some thoughts that I have:

- Right now, the native dropdown has been used for selecting the stream. I wonder if something similar to the stream selector used in the compose box is required here.
- Also, due to the native dropdown being used when the stream names get very long it looks very odd.
- When the displayed URL is not valid ("Integration URL will appear here" is shown) should the `Copy URL` button be disabled?
- Also, there is no feedback when the `Copy URL` button is clicked. (#26181) Some suggestions on what can be done about this are more than welcome.

<hr>

**Screenshots and screen captures:**

Light Theme:

<details>

<summary>Link Icon to open the modal</summary>

![image](https://github.com/sbansal1999/zulip/assets/35286603/872f34c7-e80d-4ad4-a2c4-fdd48d562e59)

</details>

<details>

<summary>Button present in Edit Bot form</summary>

![image](https://github.com/sbansal1999/zulip/assets/35286603/afa4b1a3-8d3a-4ebd-b608-e8de58254c18)

</details>

<details>

<summary>Modal for generating Integration URL</summary>

![image](https://github.com/sbansal1999/zulip/assets/35286603/00da3d8c-0fb9-42a4-9049-bc63e24ca100)

</details>

<details>

<summary>Generating Integration URL</summary>

[Screencast from 06-07-23 09:17:22 PM IST.webm](https://github.com/sbansal1999/zulip/assets/35286603/ab27228e-ef3c-4556-9179-1554a6c9e60d)



</details>

<details>

<summary>Mobile view of the modal</summary>

![image](https://github.com/sbansal1999/zulip/assets/35286603/1d9ffdba-e60f-45c1-a85b-2f125fb04dd9)

</details>

<hr>

Dark Theme:

<details>

<summary>Link Icon to open the modal</summary>

![image](https://github.com/sbansal1999/zulip/assets/35286603/0d7bddc0-d144-426d-94d3-daed35c0a12e)

</details>

<details>

<summary>Modal for generating Integration URL</summary>

![image](https://github.com/sbansal1999/zulip/assets/35286603/2cd85e8a-286e-4dfd-a1d4-0f553402cd83)


</details>

<details>

<summary>Keyboard tabbing through the modal</summary>

[Screencast from 06-07-23 09:16:21 PM IST.webm](https://github.com/sbansal1999/zulip/assets/35286603/2f6149e7-25a4-4b5e-9a0e-158224aa5aa2)


</details>

<hr>

<details>

<summary>Updated integration documentation (integration supporting event types)</summary>

![image](https://github.com/zulip/zulip/assets/35286603/56da02be-ff8a-4ae3-b9f5-ae55146f9f21)

</details>

<details>

<summary>Updated integration documentation (integration not supporting event types)</summary>

![image](https://github.com/zulip/zulip/assets/35286603/158df47f-3fad-4c6c-80ae-65b5c42c31e5)

</details>

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<hr>

Fixes part of #25976.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sbansal1999 commented:
@zulipbot add "area: bots" "area: integrations" "QA needed"

amanagr commented:
@sahil839 looks like something you should review.

sahil839 commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -0,0 +1,42 @@
+            <input
+              type="checkbox"
+              id="override-topic"
+              class="update-url"
```
I think we can use a better class here. Probably something like `integration_url_parameter`.

sbansal1999 replied:
Yeah, that class name sounds better.

sahil839 commented on web/src/settings_bots.js:
```diff
@@ -511,6 +523,103 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+        const checked = $(this).prop("checked");
+        $topic_input.toggleClass("hide", !checked);
+        $topic_name.val("");
+        update_url();
```
Since you are calling `update_url` for the `.update-url` handler, can we remove it from other places and just add that class to inputs which affect the url. I am not sure we would still need to call it from `updater` function but you can check.

sbansal1999 replied:
That is possible and would certainly make the code look cleaner.

sahil839 commented on web/src/settings_bots.js:
```diff
@@ -511,6 +523,103 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+            $override_topic.prop("disabled", false);
+            $override_topic_group.removeClass("control-label-disabled");
+        }
+    });
```
We should move all these code after `dialog_widget.launch` to be called from `post_render` argument of `dialog_widget` to make sure that handlers are registered only after the element is rendered.

sbansal1999 replied:
Makes sense!

sahil839 reviewed (commented):
Posted some comments. I tested it and looks good overall.

And yeah I agree that using the stream selector widget instead of native dropdown would be better.

sbansal1999 commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -0,0 +1,42 @@
+<div class="new-style">
+    <div class="input-group">
+        <label for="integration-name">{{t "Integration"}}</label>
+        <input type="text" id="integration-name" autocomplete="off"/>
```
Yes, that makes everything more uniform. Thanks!

sbansal1999 commented on web/src/settings_bots.js:
```diff
@@ -511,6 +523,103 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+    const $integration_url = $("#generate-integration-url-modal .integration-url");
+    const $integration_name = $("#generate-integration-url-modal #integration-name");
+    const $topic_input = $("#generate-integration-url-modal .topic-input");
+    const $topic_name = $("#generate-integration-url-modal #topic-name");
```
Agreed! The names are confusing. I will make sure to rename things and use the `.parent()` to hide things instead of creating a class for each.

sbansal1999 commented on web/styles/settings.css:
```diff
@@ -1963,3 +1967,30 @@ $option_title_width: 180px;
+            margin: auto;
+            word-wrap: break-word;
+            word-break: break-all;
+            white-space: normal;
```
Here `white-space: normal;` can be removed as well as `display: block;` as those are the defaults. Also, didn't knew about the `word-wrap` alias; that also I will fix.

sbansal1999 commented on web/src/settings_bots.js:
```diff
@@ -511,6 +523,103 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+    const $topic_input = $("#generate-integration-url-modal .topic-input");
+    const $topic_name = $("#generate-integration-url-modal #topic-name");
+    const $stream_name = $("#generate-integration-url-modal #stream-name");
+    const $override_topic = $("#generate-integration-url-modal #override-topic");
```
Making the IDs more specific seems like a good option to me.

sbansal1999 commented:
**Updates:**

- Added `modal_text_input` class to all the text inputs in the modal.
- Renamed the ID of some elements to something more specific to the modal and the thing it represented.
  Eg: `topic-name` changed to `integration-url-modal-topic-input`
- Replaced the native selector for stream selection with the dropdown widget.
- Removed some redundant styles. 
- Now the Copy URL button is disabled till the URL is invalid.
- Added a confirmation tippy that says "Copied" when the "Copy URL" button is pressed.
- Updated the documentation to show the "Generate the URL..." message in a new step. 


<details>

<summary> Screenshot of the updated widget </summary>

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/35286603/138d5c19-c542-4e36-8796-f0959e36a1a1)  | ![image](https://github.com/zulip/zulip/assets/35286603/77c3ddf8-1985-4a7d-85ae-ddd2fb1dbb55)  | 

</details>

<details>

<summary> Screenshot of the updated documentation </summary>


| Before | After |
|--------|--------|
|![image](https://github.com/zulip/zulip/assets/35286603/4bd1009a-caed-4f34-a924-800345568d3f)  | ![image](https://github.com/zulip/zulip/assets/35286603/382874cd-f53e-4bb4-b4b6-e88875c05e7c)  | 

</details>



<hr>


I wanted some feedback if this is handling long stream names correctly. 

<details>

<summary> When a stream with a very long name is selected</summary>

![image](https://github.com/zulip/zulip/assets/35286603/6c3cc08e-000f-4fb9-9bf9-d8cdf80e433f)

</details>

sahil839 reviewed (commented):
Posted few comments. LGTM otherwise.

Also, I am not sure about whether we want to format the text below stream dropdown and topic checkbox in some special way to make it looks something like a hint. We can get more feedback on this.

sbansal1999 commented on web/src/settings_bots.js:
```diff
@@ -508,6 +521,146 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+    };
+    const html_body = render_generate_integration_url_modal({
+        default_url_message,
+        streams,
```
That's a nice catch.

sbansal1999 commented on web/src/settings_bots.js:
```diff
@@ -508,6 +521,146 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+            } else {
+                stream_name = current_stream;
+                $override_topic.prop("disabled", false);
+                $override_topic_group.removeClass("control-label-disabled");
```
Yes, that's a really good idea.

sbansal1999 commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -0,0 +1,33 @@
+    </div>
+    <div class="input-group override-topic-group control-label-disabled">
+        <label class="checkbox">
+            <input type="checkbox" id="integration-url-modal-override-topic" class="integration-url-parameter" disabled/>
```
I would have loved to use short class names instead of IDs but we have a label associated with each input and hence I don't think we can removed these IDs altogether.

sahil839 replied:
Ah, yes. I forgot about the label `for` attributes. We can think of shorter IDs in that case.

sbansal1999 replied:
I can remove the `integration-url-modal` prefix from the IDs but then the IDs won't be that unique and won't convey that they are being used here. I am also not able to think of something shorter that might work.

sahil839 replied:
I think we can remove "modal" for now and name them as `integration-url-selected-stream`, etc. And for the integration input I think it would be sufficient to name it `integration-input`.

sbansal1999 replied:
Makes sense. I have pushed the changes.

sbansal1999 commented:
**Updates:**

- Removed an extra parameter that was passed to the template.
- Used `.closest(...)` to remove the `override-topic-group` class from the template.

sbansal1999 commented:
**Updates:**

- Renamed some IDs to make them a bit shorter and more readable.

sahil839 commented:
I am facing a bug where I cannot select a stream in the modal. Also, I was able to select the stream when I opened the modal for first time, but in that case the URL was not updated when selecting or changing the stream. And now after opening the modal after closing it once, I cannot select stream at all. It did work correctly before, so might be due to recent updates.
![integration-url](https://github.com/zulip/zulip/assets/35494118/35fdaddf-7baf-4815-8ba1-c04edad7ab53)

sbansal1999 commented:
**Updates:**
- Fixed the bug which caused the dropdown to not work when the modal was opened for the second time. This was done by declaring the jQuery objects inside the `tem_click_callback` function for the `dropdown_widget`. Also started a [discussion](https://chat.zulip.org/#narrow/stream/49-development-help/topic/dialog.20.20widget.20click.20callback) on why this could be happening.
- Renamed the `show_generate_url_modal` function to `show_generate_integration_url_modal`.

sbansal1999 commented:
**Updates:**
- Replaced the use of `dropdown_widget.setup()` with it's class implementation that was created in 6efcb7a.
- Changed the exit button text of the modal to "Close" instead of "Cancel". Similar to what is done in 8db9093.

@sahil839 Sorry for the delay. The bug that you identified is fixed now. Let me know if anything more is required in this one.

I think the gap between the stream dropdown widget and the "If you do not specify..." text is a bit too much. So, I wonder if I should add "-10px" top margin to the text or not.

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/35286603/d5e4e8bd-20a0-4687-bbe6-ebdd9485a37f)  | ![image](https://github.com/zulip/zulip/assets/35286603/cf7ff7e2-9b0f-4015-91dd-4e68b519a7dd)  |

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -11495,6 +11495,24 @@ paths:
+                                A `null` value means that the bot does not make use of the
+                                event types feature.
+
+                                **Changes**: New in Zulip 8.0 (feature level 193).
```
This should be updated to use the correct feature level.

sbansal1999 replied:
Nice catch!

sahil839 commented on web/styles/settings.css:
```diff
@@ -1995,3 +1999,29 @@ $option_title_width: 180px;
+        word-break: break-all;
+    }
+
+    #integration-url-selected-stream {
```
I don't think this class is used anywhere now, but I guess CSS will still be needed for long stream names.

sbansal1999 replied:
This is how it looks for long stream names which I think is fine. 

![image](https://github.com/zulip/zulip/assets/35286603/3134ebeb-ae55-4866-aaa0-56da99788e4b)

sahil839 replied:
The other dropdowns also look like this, but I think keeping the height of button same and change the text to something like `very long very long ......` would be better. We can use tooltips to show the full name. I think we can discuss this in #design on CZO.

sahil839 reviewed (commented):
Posted a couple of comments. Also, I noticed that the bottom border disappears for the url box when the url is long.
![incoming-bot-url](https://github.com/zulip/zulip/assets/35494118/1606108b-f7e3-4e6d-93d9-e73cbbd3b1c5)

sahil839 commented:
> I think the gap between the stream dropdown widget and the "If you do not specify..." text is a bit too much. So, I wonder if I should add "-10px" top margin to the text or not.
Yeah, we can decrease that gap. Also I think we can make the gap between the stream paragraph and the `Override topic` input consitent with others, by using `input-group` class on the parent div of stream-widget.

sbansal1999 commented:
**Updates:**
- Added horizontal padding to the dialog widget modal content section to avoid overlaps.
- Changed the appearance of the dropdown widget when long stream names are selected; now the stream name is cut off and an ellipsis is shown to denote that.
- Decreased the gap between the stream not specified notice and the dropdown.
- Made gapping between input groups more consistent as suggested in [this](https://github.com/zulip/zulip/pull/26201#issuecomment-1657453309) comment.

sbansal1999 commented:
> We can use tooltips to show the full name. I think we can discuss this in #design on CZO.

Started a discussion about this here: https://chat.zulip.org/#narrow/stream/101-design/topic/long.20stream.20names.20in.20dropdown/near/1618927

sahil839 commented:
Looks good to me. We can mark it for integration review once we come on conclusion in the design stream.

sbansal1999 commented:
**Updates:**

- Changed the width of the stream selector to 75% making room for long stream names.
- Aligned the text shown in the stream selector to the left to make it more readable.

<details>

<summary> Current View when a long stream name is selected </summary>

![image](https://github.com/zulip/zulip/assets/35286603/bd15c6a8-a302-4a5c-a5b1-1cd803933bc2)
 
</details>

sbansal1999 commented:
**Updates:**

-Rebased the PR and fixed the import order CI error.

sbansal1999 commented:
**Updates:**
- Modified the behavior in case of long stream names according to [this](https://chat.zulip.org/#narrow/stream/101-design/topic/long.20stream.20names.20in.20dropdown/near/1619353) comment.


<details>

<summary> Long stream names </summary>

![image](https://github.com/zulip/zulip/assets/35286603/7ca7280f-55d6-48a5-afda-810a04c318bb)


</details>

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

sahil839 commented:
Looks good to me. You would need to rebase the PR again and then mark it ready for review by Tim.

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

timabbott commented on zerver/lib/integrations.py:
```diff
@@ -830,3 +830,12 @@ def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
 }
+
+
+def get_all_event_types(integration: Integration) -> Optional[List[str]]:
```
`all_event_types_for_integration` would be a nicer name that's readable without having to check where this imported from.

sbansal1999 replied:
Agreed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11518,6 +11518,24 @@ paths:
+                            display_name:
+                              type: string
+                              description: |
+                                The name of the bot that is displayed to the users.
```
This documentation is super confusing to someone who doesn't know about this system. This line should clearly explain how this is different from `name`. E.g.  (please check this is correct, I'm going off memory):

> A machine-readable unique name uniquely identifying an official Zulip integration, all-lower-case without spaces.

and 

> A human-readable display name identifying an official Zulip integration that this bot implements.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11518,6 +11518,24 @@ paths:
                               type: string
                               description: |
                                 The name of the bot.
+                            display_name:
```
Is `null` possible here?

sbansal1999 replied:
No, it cannot be `null`. 

The following lines of code in the `__init__` function of the `Integration` class which is the parent class of `WebhookIntegration` make sure that each integration has a `display_name`:

https://github.com/zulip/zulip/blob/d205850d54f73d648488bc1dbd37cca2dfdef272/zerver/lib/integrations.py#L104-L106

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11518,6 +11518,24 @@ paths:
+                                type: string
+                              nullable: true
+                              description: |
+                                The event types that are supported by the bot.
```
"For incoming webhook integrations that support the Zulip server filtering incoming events, the list of event types supported by the integration. A `null` value will be present if this is not an incoming webhook integration."

sbansal1999 replied:
Yeah, this describes the whole thing precisely.

timabbott commented on zerver/webhooks/deskdotcom/doc.md:
```diff
@@ -1,17 +1,17 @@
-1.  {!create-stream.md!}
+1. {!create-stream.md!}
```
@sbansal1999 can you explain what you're achieving with these spacing edits? It might be nice to pull this out to a separate prep PR and so @laurynmm can review it without needing to look at the rest of the changes here.

sbansal1999 replied:
Opened #26484 .

timabbott commented on web/styles/modal.css:
```diff
@@ -86,7 +86,7 @@
     font-size: 1rem;
     overflow-y: auto;
-    padding: 0 24px;
+    padding: 2px 24px;
```
Can you explain this change more, maybe with before/after screenshots? Might be good to just do this in a separate prep PR since this PR thread is already very long.

sbansal1999 replied:
Opened #26500.

timabbott commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -0,0 +1,33 @@
+              widget_name="integration-url-stream"
+              label=(t "Stream")}}
+        </div>
+        <p id="stream-not-specified-notice">{{t "If you do not specify a stream, the bot will send notifications via PMs to the creator of the bot."}}</p>
```
Having this block of text inline feels awkward. I feel like ideally we'd just have the default choice in this stream picker be "Direct message to me" and adjust the label a bit so that feels natural, e.g. "Where to send notifications"

sbansal1999 replied:
Yes, that sounds right. This should also clear things up in the form itself.

timabbott commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -0,0 +1,33 @@
+            <span></span>
+        </label>
+        <label class="inline" for="integration-url-override-topic">
+            {{t "Override topic for notifications?"}}
```
How about "Send all notifications to a single topic" for this label -- and then we can remove the explanatory paragraph below. (And maybe as a `help_link_widget` with help text "By default, most integrations will send notifications to multiple topics in a thoughtful way.")

sbansal1999 replied:
Hmm, that sounds like a good idea. 

I wonder where should the `help_link_widget` link to. Maybe I can add the following to the first bullet point on [this](https://zulip.com/api/integrations-overview#set-up-an-existing-integration) page:

> After successful setup, notifications will be sent to the specified stream and most integrations will send them to multiple topics in a thoughtful way.

Or do we require a separate page for this? I feel that would be too much for just this one piece of information.

timabbott commented:
Thanks for all the work on this @sbansal1999! I posted a few quick comments that should help us integrate a bunch of the prep commits to shrink this a bunch and give you some ideas for improving the design of the form.

sbansal1999 commented:
**Updates**:
- Changed the first option from "Select stream" to "Direct message to me" resulting in some text being removed.
- Made use of `stream_data.subscribed_subs()` to get the stream data to make the dropdown widget a bit more good-looking.

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/35286603/85ce7ef3-2526-4d5f-a846-0822e351477d)  | ![image](https://github.com/zulip/zulip/assets/35286603/70541b9c-6df8-4306-a304-da3d140798f6)  | 


- Renamed a couple of labels resulting in the removal of unnecessary information.

Also, I have left a [question](https://github.com/zulip/zulip/pull/26201#discussion_r1295045415) about creating a `help_link_widget` that would be associated with the topic input field.

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

sbansal1999 commented:
**Updates:**

- Rebased the PR and fixed the merge conflicts.

sbansal1999 commented:
@timabbott Bumping this one in case you might have missed this.

alya commented:
I think this got delayed on my review -- sorry about that!

alya commented:
On the instructions on the integrations pages, I think we'll want to add a link from "profile card" to a help page with instructions on how to find the bot profile card. I filed an issue for getting that documented:

- #26796

alya commented:
I have a couple of minor UI tweaks to suggest... Having chatted with @timabbott , I'll file those as a separate follow-up issue. That way, we can merge the settings parts of this PR, make those tweaks, and then merge the `/integrations` changes.

alya commented:
Should we validate to make sure the topic name is not longer than allowed? I'm not sure what happens if it's too long.

timabbott commented:
For the topic length part, I added this:

``` diff
diff --git a/web/src/settings_bots.js b/web/src/settings_bots.js
index 781d07939a..0856c06877 100644
--- a/web/src/settings_bots.js
+++ b/web/src/settings_bots.js
@@ -529,6 +529,7 @@ export function show_generate_integration_url_modal(api_key) {
     };
     const html_body = render_generate_integration_url_modal({
         default_url_message,
+        max_topic_length: page_params.max_topic_length,
     });
 
     function generate_integration_url_post_render() {
diff --git a/web/templates/settings/generate_integration_url_modal.hbs b/web/templates/settings/generate_integration_url_modal.hbs
index 3797199cd5..432ecb4975 100644
--- a/web/templates/settings/generate_integration_url_modal.hbs
+++ b/web/templates/settings/generate_integration_url_modal.hbs
@@ -21,7 +21,7 @@
     </div>
     <div class="input-group hide">
         <label for="integration-url-topic-input">{{t "Topic"}}</label>
-        <input type="text" id="integration-url-topic-input" class="modal_text_input integration-url-parameter" />
+        <input type="text" id="integration-url-topic-input" class="modal_text_input integration-url-parameter" maxlength="{{ max_topic_length }}" />
     </div>
     <hr />
     <p class="integration-url-header">{{t "URL for your integration"}}</p>
```

timabbott commented:
I extracted the part that doesn't update the /integrations documentation as 9cde3abd4de68145071b0ac1041d3418e82784e1 and merged it. The remaining commit just updates the /integrations/ documentation; we can merge it once we've happy with some details that will be tracked in separate mini issues.

laurynmm commented:
**Updates**:
- Rebased changes to update for changes to Slack integration documentation updates.
- Added a commit to update the API and contributor documentation to replace references to `create-bot-construct-url` with the new `generate-integration-url` macro.

alya commented:
@laurynmm Should this have the "integration review" label now?

laurynmm commented:
Yes, adding the integration review label now.

timabbott commented:
Merged, thanks for all the work on this @sbansal1999 and @laurynmm!
